streambed-machine/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+use async_stream::stream;
+use futures_util::{Stream, StreamExt};
+use rand::thread_rng;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{marker::PhantomData, pin::Pin, vec::Vec};
+use streambed::{
+    commit_log::{Offset, ProducerRecord, Subscription, Topic},
+    decrypt_buf_with_secret, encrypt_struct_with_secret, get_secret_value,
+    secret_store::SecretStore,
+};
+
+pub use streambed::commit_log::{CommitLog, ProducerError};
+
+/// Wraps a `CommitLog` and specializes it for a specific payload type.
+/// This adds the type, topic and the encoding and encryption scheme.
+#[derive(Debug)]
+pub struct LogAdapter<L, C, A> {
+    commit_log: L,
+    codec: C,
+    topic: Topic,
+    marker: PhantomData<A>,
+}
+
+/// Provides a method on `CommitLog` to specialize it for a payload type.
+pub trait CommitLogExt
+where
+    Self: CommitLog + Sized,
+{
+    /// Specialize this commit log for items of type `A`
+    /// The topic and group names are given and a `Codec`
+    /// for encoding and decoding values of type `A`.
+    fn adapt<A>(
+        self,
+        topic: impl Into<Topic>,
+        codec: impl Codec<A>,
+    ) -> LogAdapter<Self, impl Codec<A>, A> {
+        LogAdapter {
+            commit_log: self,
+            codec,
+            topic: topic.into(),
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<L> CommitLogExt for L where L: CommitLog {}
+
+impl<L, C, A> LogAdapter<L, C, A>
+where
+    L: CommitLog,
+    C: Codec<A>,
+    A: 'static,
+{
+    /// Send one item to the underlying commit log.
+    pub async fn produce(&self, item: A) -> Result<Offset, ProducerError> {
+        let topic = self.topic.clone();
+
+        if let Some(value) = self.codec.encode(item) {
+            self.commit_log
+                .produce(ProducerRecord {
+                    topic,
+                    headers: Vec::new(),
+                    timestamp: None,
+                    key: 0,
+                    value,
+                    partition: 0,
+                })
+                .await
+                .map(|r| r.offset)
+        } else {
+            Err(ProducerError::CannotProduce)
+        }
+    }
+
+    /// Return an async stream of items representing the
+    /// history up to the time of the call.
+    #[allow(clippy::needless_lifetimes)]
+    pub async fn history<'a>(&'a self) -> Pin<Box<impl Stream<Item = A> + 'a>> {
+        let last_offset = self
+            .commit_log
+            .offsets(self.topic.clone(), 0)
+            .await
+            .map(|lo| lo.end_offset);
+        let subscriptions = Vec::from([Subscription {
+            topic: self.topic.clone(),
+        }]);
+
+        let mut records =
+            self.commit_log
+                .scoped_subscribe("EDFSM", Vec::new(), subscriptions, None);
```

Comment:
I don't see this as a good idea. Perhaps we should have a wider conversation about the influence of Kafka on our API though.

Reply:
I can make 'group' an argument again.   But I suppose it depends on how general we consider `LogAdapter` to be.  In the known applications of `LogAdapter`  consumer group is unused.    If consumer group names are important in some other environment then would LogAdapter be used?  Probably be using the CommitLog API directly.